### PR TITLE
[WIP] WebGL shader material and data texture

### PR DESCRIPTION
**Pull request by izaid**
This extends the WebGL interface of ThreeJS.jl by adding Julia functions for ShaderMaterial and DataTexture from ThreeJS. This allows pretty much arbitrary WebGL shaders to be attached to meshes.

It requires some changes in Patchwork.jl and Escher.jl, which have yet to be merged there.

### CODE REVIEW

File: assets/bower_components/three-js/three-js.html [hunk 1]
```diff
@@ -1241,6 +1241,152 @@
...
+        Polymer.dom(this).parentNode.addMaterial(this.object);
+      },
+      attached: function () {
+        this.async(function (){
+            this.newMaterial();
+        });
+      }
+    });
+
+  PropertyHook.setupHook("three-js-shader-material", "uniforms",
```
Comment (rohitvarkey) [inline assets/bower_components/three-js/three-js.html:1383]: Can you explain what this does?

Comment (izaid) [reply assets/bower_components/three-js/three-js.html:1383]: Yeah, sure. The PropertyHooks are a new feature of Patchwork.jl. They allow for nested Julia dictionaries to be serialized to JavaScript, and efficiently updated with Signals. We need that to allow arbitrary WebGL shaders, which should support arbitrary number of uniforms and defines as per http://threejs.org/docs/api/materials/ShaderMaterial.html.

For various reasons, using properties directly wasn't possible here. I'm going to cc @shashi, who made the changes in Patchwork.jl and Escher.jl to make this work, and who decided on the hooks approach.

**Comment by rohitvarkey** [comment]
I have actually seperated out the three-js elements to this repository - https://github.com/rohitvarkey/three-js. It'd be great if you can make a PR for just the Polymer parts there, and do a `bower update` and then update the elements here.

Otherwise, this looks good so far. Thanks @izaid for the effort!

Also, can you push the `Escher` and `Patchwork` changes to a branch somewhere so I can test this?

**Comment by izaid** [comment]
Will do!

You can get the Patchwork.jl repo from https://github.com/izaid/Patchwork.jl/tree/hooks and the Escher.jl repo from https://github.com/izaid/Escher.jl/tree/nested-props

**Comment by shashi** [comment]
Nice!

@rohitvarkey would this work in Jupyter? The `escher-prop-hook` thing comes from `Escher/assets`...

File: assets/bower_components/three-js/three-js.html (hunk 1 above)
Comment (shashi) [reply assets/bower_components/three-js/three-js.html:1383]: I think @rohitvarkey knows that Polymer cannot listen to changes to `elem.prop.subProp`, the `hooks` thing lets you do that.

**Comment by izaid** [comment]
@rohitvarkey Can I just confirm that you are okay with the fact that this works via Escher (as @shashi mentioned) and thus we're not sure if it will work in Jupyter? When you confirm that, I'll go ahead and reopen this PR with the bower update separate.

**Comment by rohitvarkey** [comment]
> would this work in Jupyter? The escher-prop-hook thing comes from Escher/assets

Nope. It wouldn't work. Unless we figure out how to serve those files in Jupyter.

> Can I just confirm that you are okay with the fact that this works via Escher

I'm okay with this.

**Comment by izaid** [comment]
Alright, I'm closing this and will reopen two separate PRs.

**Pull request closed without merging**